jbpm-flow-builder/src/main/java/org/jbpm/process/builder/dialect/mvel/MVELAssignmentAction.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jbpm.process.builder.dialect.mvel;
+
+import java.util.function.BiFunction;
+import java.util.regex.Matcher;
+
+import org.drools.mvel.MVELSafeHelper;
+import org.jbpm.process.instance.impl.AssignmentAction;
+import org.jbpm.process.instance.impl.AssignmentProducer;
+import org.jbpm.util.PatternConstants;
+import org.jbpm.workflow.core.node.Assignment;
+import org.jbpm.workflow.instance.impl.NodeInstanceResolverFactory;
+import org.kie.api.runtime.process.NodeInstance;
+import org.kie.api.runtime.process.ProcessContext;
+
+public class MVELAssignmentAction implements AssignmentAction {
+
+    private String to;
+    private String from;
+    private String srcExpr;
+    private String targetExpr;
+    private AssignmentProducer producer;
+    private BiFunction<ProcessContext, NodeInstance, Object> src;
+    private BiFunction<ProcessContext, NodeInstance, Object> target;
+
+    public MVELAssignmentAction(Assignment assignment, String sourceExpr, String targetExpr,
+                                BiFunction<ProcessContext, NodeInstance, Object> source,
+                                BiFunction<ProcessContext, NodeInstance, Object> target, AssignmentProducer producer) {
+
+        Matcher fromMatcher = PatternConstants.PARAMETER_MATCHER.matcher(assignment.getFrom());
+        Matcher toMatcher = PatternConstants.PARAMETER_MATCHER.matcher(assignment.getTo());
+
+        this.from = fromMatcher.find() ? fromMatcher.group(1) : assignment.getFrom();
+        this.to = toMatcher.find() ? toMatcher.group(1) : assignment.getTo();
+        this.src = source;
+        this.target = target;
+        this.srcExpr = sourceExpr;
+        this.targetExpr = targetExpr;
+        this.producer = producer;
+    }
+
+    @Override
+    public void execute(NodeInstance nodeInstance, ProcessContext context) throws Exception {
+        Object targetObject = this.target.apply(context, nodeInstance);
+        Object srcObject = this.src.apply(context, nodeInstance);
+        NodeInstanceResolverFactory resolver = new NodeInstanceResolverFactory(
+                (org.jbpm.workflow.instance.NodeInstance) nodeInstance);
+
+        // if just evaluating, not assignment
+        if (targetObject == null || notEvalTarget()) {
+            producer.accept(context, nodeInstance, notEvalSrc() ? srcObject : MVELSafeHelper.getEvaluator().eval(
+                    from, srcObject, resolver));
+        } else {
+            resolver.addExtraParameter(srcExpr, srcObject);
+            resolver.addExtraParameter(targetExpr, targetObject);
+            MVELSafeHelper.getEvaluator().eval(ensureLocated(targetExpr, to).concat("=").concat(ensureLocated(srcExpr,
+                    from)), resolver);
+        }
+    }
```

Comment:
The concept of "." as "this" is does not feel very standard.
The concept of expression [] for arrays is as well very strange without setting the name of the array.

To make things a bit more standard we should remove the logic of ensureLocated as you are trying to infer things. 
We can can set the extra parameters as source/target for instance as a pointer to the object. or even something like 
@source[3].myprop = @target.myprop. This will avoid the problem of having node output express with the same name as target. (Edge case but could happen)

This should leave trail in the audit table so I would suggest to add a test about that to evaluate the source expression
target expression
the call producer

in three different steps.

Please also remove the need of #{} if the UI is no setting the language you have already more than enough check to work that properly.

Reply:
this keyword introduced to cope with the egde case, if expression has same name than target, this usage is mandatory, otherwise is optional

Reply:
examples edited to not use #{} if language is specified in expression, otherwise it needs to be specified to identify the expression as mvel

Reply:
this prefix is now used in list assigment example